Update subkey to v2.0.1 and extend examples (#992)
substrate-developer-hub/substrate-developer-hub.github.io: docs/knowledgebase/integrate/subkey.md

[within Substrate itself](https://github.com/paritytech/substrate/tree/master/bin/utils/subkey). 
Its main feature is generating and inspecting key pairs, currently supporting these scheme:

- [sr25519](https://wiki.polkadot.network/docs/en/learn-cryptography)
- [ed25519](https://en.wikipedia.org/wiki/EdDSA#Ed25519)
- [secp256k1](https://en.bitcoin.it/wiki/Secp256k1)

All keys in substrate based networks
([like polkadot](https://wiki.polkadot.network/docs/en/learn-accounts#address-format))
use the 
[SS58 address encoding format](https://github.com/paritytech/substrate/wiki/External-Address-Format-(SS58))
that is the primary user-facing way to address and interact with keys.

Subkey also allows restoring keys from mnemonics and raw seeds, signing and verifying signatures
on a message, even for encoded transactions.

## Installation

### Install with Cargo

You will need to have the Substrate build dependencies to install subkey.
Use the following two commands to install the dependencies and Subkey, respectively:

_Command:_
```bash
# Don't install substrate and subkey binary, just get deps needed with `--fast` flag
curl https://getsubstrate.io -sSf | bash -s -- --fast 

> # Install only `subkey`, at specific version (tmp pulls files, builds, installs bin)

Author: ```suggestion
# Install only `subkey`, at a specific version
```
this is what the command actually does. It can be removed to avoid confusion

  to write down your key by hand.
- **Secret Seed** (aka "Private Key" or "Raw Seed") - The minimum necessary information to restore
  the key pair. All other information is calculated from the seed.
- **Public Key (hex)** - The public half of the cryptographic key pair in hexadecimal.
- **Public Key (SS58)** - The public half of the cryptographic key pair in SS58 encoding.
- **Account ID** - Alias for the Public Key in hexadecimal. 
- **SS58 Address** (aka "Public Address") - An SS58-encoded address based on the public key.

> Currently Subkey supports the following cryptographic key pairs and signing algorithms:

Reviewer: Maybe this should go because it's stated already at the top.

Reviewer: If yes, then I suggest to include links for each:
- sr25519:  Schorr signatures on the Ristretto group
- ed25519: ed25519
- secp256k1: ECDSA signatures on secp256k1

Author: included in the first instance

```

## Specify Network Address Format

Subkey encodes the address depending on the network. You can use the `--network` flag to change
this. See `subkey <inspect/generate> -n` for supported networks.

Let's say you want to use the **same private key** on the Kusama network. Use `--network` to get
your **address** formatted for Kusama. Notice that the public key is the same, but the address has
a different format. Here we use the [example seed](#example-seed):

_Command:_
```bash
subkey inspect --network kusama "caution juice atom organ advance problem want pledge someone senior holiday very"
```
_Output:_
```text
Secret phrase `caution juice atom organ advance problem want pledge someone senior holiday very` is account:
  Secret seed:       0xc8fa03532fb22ee1f7f6908b9c02b4e72483f0dbd66e4cd456b8f34c6230b849
  Public key (hex):  0xd6a3105d6768e956e9e5d41050ac29843f98561410d3a47f9dd5b3b227ab8746
  Public key (SS58): HRkCrbmke2XeabJ5fxJdgXWpBRPkXWfWHY8eTeCKwDdf4k6
  Account ID:        0xd6a3105d6768e956e9e5d41050ac29843f98561410d3a47f9dd5b3b227ab8746
  SS58 Address:      HRkCrbmke2XeabJ5fxJdgXWpBRPkXWfWHY8eTeCKwDdf4k6
```

## HD Key Derivation

> > REMEMBER: **Passwords and derivation paths are just as important to backup as your seed itself!**

Reviewer: ```suggestion
> **Remember that passwords and derivation paths are just as important to backup as the seed itself!**
```

  Public key (SS58): 5FRjccB7s9fbMu4pwQhYng2quQnKYkCHXRUCMBRwL7Pzj8FX
  Account ID:        0x94bec5342b23d74b8736c74ab1bf311182c62510d220313a424e63ea57172855
  SS58 Address:      5FRjccB7s9fbMu4pwQhYng2quQnKYkCHXRUCMBRwL7Pzj8FX
```

Recall the _root_ SS58 address from the [example seed](#example-seed) is,
`5Gv8YYFu8H1btvmrJy9FjjAWfb99wrhV3uhPFoNEr918utyR`. We can use that _public address_ to derive the 
same child address as above using the _seed_.

_Command:_
```bash
subkey inspect "5Gv8YYFu8H1btvmrJy9FjjAWfb99wrhV3uhPFoNEr918utyR/polkadot/0"
```
_Output:_
```text
Public Key URI `5Gv8YYFu8H1btvmrJy9FjjAWfb99wrhV3uhPFoNEr918utyR/polkadot/0` is account:
  Network ID/version: substrate
  Public key (hex):   0x94bec5342b23d74b8736c74ab1bf311182c62510d220313a424e63ea57172855
  Public key (SS58):  5FRjccB7s9fbMu4pwQhYng2quQnKYkCHXRUCMBRwL7Pzj8FX
  Account ID:         0x94bec5342b23d74b8736c74ab1bf311182c62510d220313a424e63ea57172855
  SS58 Address:       5FRjccB7s9fbMu4pwQhYng2quQnKYkCHXRUCMBRwL7Pzj8FX
```

Note that the two addresses here match. This is not the case in hard key derivation!

## Password Protected Keys

> > REMEMBER: **Passwords and derivation paths are just as important to backup as your seed itself!**

Reviewer: ```suggestion
> **Remember that passwords and derivation paths are just as important to backup as the seed itself!**
```

  Account ID:        0xdef8f78b123475265815b65a7c55e105e1ab185f4969954f68d92b7bb67a1045
  SS58 Address:      5H74SqH1iQCWh5Gumyghh1WJMcmM6TdBHYSK7mKVJbv9NuSK
```

_Command (password with `///`):_
```bash
subkey inspect "caution juice atom organ advance problem want pledge someone senior holiday very///pencil laptop kitchen cutter"
```
_Output (password with `///`):_
```text
Secret Key URI `caution juice atom organ advance problem want pledge someone senior holiday very///pencil laptop kitchen cutter` is account:
  Secret seed:       0xdfc5d5d5235a37fdc907ee1cb720299f96aeb02f9c7c2fcad7ee8c7bfbd2a4db
  Public key (hex):  0xdef8f78b123475265815b65a7c55e105e1ab185f4969954f68d92b7bb67a1045
  Public key (SS58): 5H74SqH1iQCWh5Gumyghh1WJMcmM6TdBHYSK7mKVJbv9NuSK
  Account ID:        0xdef8f78b123475265815b65a7c55e105e1ab185f4969954f68d92b7bb67a1045
  SS58 Address:      5H74SqH1iQCWh5Gumyghh1WJMcmM6TdBHYSK7mKVJbv9NuSK
```
Take note that the `--password` flag _does not_ report the password used, but the `///` _does_. 
The output is identical otherwise.

## Combine Derivation Paths & Passwords

### IMPORTANT PASSWORD AND DERIVATION NOTE

> > Note that the "Secret seed" is the same and _not_ password protected, it can still recover _an_
> > account - but the key pair derived _is not the same_ account as recovered with _any_ password!
> > You can see this in the next section demonstrated.

Reviewer: ```suggestion
> Note that the "secret seed" is the same and _is not_ password protected. Although it can still recover _an_
> account, the key pair that's derived _is not the same_ account as recovered with _any_ password!
> This is demonstrated in the next section.
```

> 

> > **The "Secret phrase" is not sufficient to recover the account without the password!**

Reviewer: Isn't this repeating the above?

Author: This IMHO needs to be as stressed as possible, do you think this is too much here?

> Keep this password secure, as without it your key pair cannot be recovered!

You can mix and match hard and soft key paths (although it doesn't make much sense to have hard
paths as children of soft paths). For example:

_Command:_
```bash
subkey inspect "caution juice atom organ advance problem want pledge someone senior holiday very//polkadot/0"
```
_Output:_
```text
Secret Key URI `caution juice atom organ advance problem want pledge someone senior holiday very//polkadot/0` is account:
  Secret seed:       n/a
  Public key (hex):  0xd6f066a07b3ebb1572c00506c9dc86a64db63779e7791e7d37247f59e4bffc14
  Public key (SS58): 5GvXX2NpiR8qoeTUbdzTCmMtZgGrZxRoCJFEvfL1iaDjTFKb
  Account ID:        0xd6f066a07b3ebb1572c00506c9dc86a64db63779e7791e7d37247f59e4bffc14
  SS58 Address:      5GvXX2NpiR8qoeTUbdzTCmMtZgGrZxRoCJFEvfL1iaDjTFKb
```

The first level (`//polkadot`) is **hard-derived**, while the leaf (`/0`) is **soft-derived**.

To use key derivation with a password-protected key, add your password to the end:

_Command:_
```bash

> # mnemonic phrase plus derivation path plus password

Reviewer: ```suggestion
# Mnemonic phrase + derivation path + password
```

  Secret seed:       n/a
  Public key (hex):  0xcaf1f5ec14b507b5c365c6528cc06de74a5615274694a0c895ed4109c0ff0d32
  Public key (SS58): 5GeoQa3nkeNmzZSfgBFuK3BkAggnTHcX3S1j94sffJYYphrP
  Account ID:        0xcaf1f5ec14b507b5c365c6528cc06de74a5615274694a0c895ed4109c0ff0d32
  SS58 Address:      5GeoQa3nkeNmzZSfgBFuK3BkAggnTHcX3S1j94sffJYYphrP
```

Notice that for **soft-derived** keys, you can use the hard-derived (with optional password)
**parent's public address** to derive new public addresses:

_Command:_
```bash
# A soft-derived SS58-address for a *public address* with a hidden seed, hard derivation path, and password
subkey inspect "5GsbzysSK8TKahXBC7FpS2myx3nWehMyYU7q8CLrzZCjpKbM/0"
```
_Output:_
```text
Public Key URI `5GsbzysSK8TKahXBC7FpS2myx3nWehMyYU7q8CLrzZCjpKbM/0` is account:
  Network ID/version: substrate
  Public key (hex):   0xcaf1f5ec14b507b5c365c6528cc06de74a5615274694a0c895ed4109c0ff0d32
  Public key (SS58):  5GeoQa3nkeNmzZSfgBFuK3BkAggnTHcX3S1j94sffJYYphrP
  Account ID:         0xcaf1f5ec14b507b5c365c6528cc06de74a5615274694a0c895ed4109c0ff0d32
  SS58 Address:       5GeoQa3nkeNmzZSfgBFuK3BkAggnTHcX3S1j94sffJYYphrP
```

> Notice that the "SS58-address plus derivation path" produces the same address as the "mnemonic
> phrase plus derivation path plus password." As such, you can reveal your parent public address

Reviewer: ```suggestion
Notice that the "SS58-address + derivation path" produces the same address as the "mnemonic
phrase + derivation path + password." As such, you can reveal your parent public address
```

```
_Output:_
```text
12D3KooWAhzqC4xstZKkAVYge2Q9tAMcnFhNEabLejRNVoQWLHHC
```

The peer ID is displayed on screen and the actual key is saved in the `<output-file>`.

> <!-- FIXME! The `insert` command is _only_ for the key sub-command embedded in nodes. Not for subkey - we need to point this out... -->

Reviewer: Not sure why there's a FIXME here ?

Author: see #993
